packages/router/router.ts

```diff
@@ -3284,6 +3299,52 @@ export function createRouter(init: RouterInit): Router {
     );
   }
 
+  // Standard implementation - patch children into a parent and `updateState`
+  function patchRoutes(
+    routeId: string,
+    children: AgnosticDataRouteObject[]
+  ): void;
+  // Batch implementation - perform multiple patches followed by a single `updateState`
+  function patchRoutes(
+    cb: (
+      patch: (routeId: string, children: AgnosticDataRouteObject[]) => void
+    ) => void
+  ): void;
+  function patchRoutes(
+    routeId:
+      | string
+      | ((
+          patch: (routeId: string, children: AgnosticDataRouteObject[]) => void
+        ) => void),
+    children?: AgnosticDataRouteObject[]
+  ): void {
+    let isNonHMR = inFlightDataRoutes == null;
+    let routesToUse = inFlightDataRoutes || dataRoutes;
+    if (typeof routeId === "function") {
+      routeId((r, c) =>
+        patchRoutesImpl(r, c, routesToUse, manifest, mapRouteProperties)
+      );
```

Comment:
If this is the batch API usage - apply all patches then `updateState` a single time below